### PR TITLE
Implement RECOVERY run feature.

### PR DESCRIPTION
**Pull request by cyulysses-corp**

### CODE REVIEW

File: src/javatests/com/google/cloud/bigquery/dwhassessment/extractiontool/executor/SaveCheckerImplTest.java [hunk 1]
```diff
@@ -45,6 +46,24 @@ public void setUp() throws IOException {
         new SaveCheckerImpl(ImmutableMap.of(SCRIPT_NAME, ImmutableList.of("testTimestampColumn")));
   }
 
+  @Test
+  public void findFinishedRecords_onlyReturnsCorrectOnes() throws IOException {
```
Comment (frsv) [inline src/javatests/com/google/cloud/bigquery/dwhassessment/extractiontool/executor/SaveCheckerImplTest.java:50]: Would be nice to add more tests covering edge cases:
* Empty input
* No matching files
* All files matches

File: src/javatests/com/google/cloud/bigquery/dwhassessment/extractiontool/executor/ExtractExecutorImplTest.java [hunk 2]
```diff
@@ -247,6 +249,95 @@ public void run_incrementalMode_success() throws Exception {
...
+    verify(scriptManager)
+        .executeScript(
+            any(Connection.class),
+            /*dryRun=*/ eq(false),
+            any(SqlTemplateRenderer.class),
+            /*scriptName=*/ eq("script_no_record"),
+            eq(dataEntityManager),
+            eq(0),
+            eq(0));
+    verifyNoMoreInteractions(scriptManager);
```
Comment (frsv) [inline src/javatests/com/google/cloud/bigquery/dwhassessment/extractiontool/executor/ExtractExecutorImplTest.java:287]: nit: such an assertion may have less sense if do not expect calls to other methods of the class – e.g.:

```
bool fooBar = scriptManager.foo();
if (fooBar) {
  scriptManager.bar();
}
```

Otherwise, single `verify` statement is enough and `verifyNoMoreInteractions` looks slightly over-defensive

File: src/java/com/google/cloud/bigquery/dwhassessment/extractiontool/executor/ExtractExecutorImpl.java [hunk 3]
```diff
@@ -126,19 +126,38 @@ private static void maybeAddTimeRange(
 
   @Override
   public int run(Arguments arguments) throws SQLException, IOException {
+    Preconditions.checkArgument(
```
Comment (frsv) [inline src/java/com/google/cloud/bigquery/dwhassessment/extractiontool/executor/ExtractExecutorImpl.java:129]: I may missed that, but I didn't notice tests for this case. Could you add it?

File: src/java/com/google/cloud/bigquery/dwhassessment/extractiontool/subcommand/ExtractSubcommand.java [hunk 4]
```diff
@@ -348,31 +341,19 @@ private ExtractExecutor.Arguments getValidatedArguments() {
...
-      if (!Files.isDirectory(path)) {
-        throw new ParameterException(
-            spec.commandLine(),
-            String.format(
-                "--prev-run-path must specify a directory, but '%s' is not a directory.", path));
-      }
-      argumentsBuilder.setPrevRunPath(path);
+    switch (mode) {
+      case INCREMENTAL:
+        validateAndSetPrevRunPathIncrementalMode();
```
Comment (frsv) [inline src/java/com/google/cloud/bigquery/dwhassessment/extractiontool/subcommand/ExtractSubcommand.java:346]: `getValidatedArguments` becomes a majority of useful code existing in this class. Also, most of the validation is repetitive (check ending, check if path or parent is a directory).

My suggestion is to move the validation logic to a separate class with a clear control flow (e.g. based on the parameter name), maybe in a separate CL.

File: src/java/com/google/cloud/bigquery/dwhassessment/extractiontool/subcommand/ExtractSubcommand.java [hunk 5]
```diff
@@ -408,6 +389,64 @@ private ExtractExecutor.Arguments getValidatedArguments() {
...
+    String pathString = prevRunPathString == null ? outputPathString : prevRunPathString;
+    if (pathString.endsWith(".zip")) {
+      throw new ParameterException(
+          spec.commandLine(), "Recovery mode is not supported for zipped records, yet.");
+    }
+    Path path = Paths.get(pathString);
+    if (!Files.isDirectory(path)) {
+      throw new ParameterException(
+          spec.commandLine(),
+          String.format("The path '%s' you specified is not a directory.", path));
```
Comment (frsv) [inline src/java/com/google/cloud/bigquery/dwhassessment/extractiontool/subcommand/ExtractSubcommand.java:445]: Could you add a test to cover this?

File: src/java/com/google/cloud/bigquery/dwhassessment/extractiontool/subcommand/ExtractSubcommand.java [hunk 6]
```diff
@@ -408,6 +389,64 @@ private ExtractExecutor.Arguments getValidatedArguments() {
...
+      throw new ParameterException(
+          spec.commandLine(),
+          String.format(
+              "--prev-run-path must specify a directory, but '%s' is not a directory.", path));
+    }
+    argumentsBuilder.setPrevRunPath(path);
+  }
+
+  private void validateAndSetPrevRunPathRecoveryMode() {
+    String pathString = prevRunPathString == null ? outputPathString : prevRunPathString;
```
Comment (frsv) [inline src/java/com/google/cloud/bigquery/dwhassessment/extractiontool/subcommand/ExtractSubcommand.java:436]: Could you add a test to cover a case when previous path is specified?

File: src/java/com/google/cloud/bigquery/dwhassessment/extractiontool/subcommand/ExtractSubcommand.java [hunk 7]
```diff
@@ -408,6 +389,64 @@ private ExtractExecutor.Arguments getValidatedArguments() {
     return arguments;
   }
 
+  private void setOutputPath() {
+    Path path = Paths.get(outputPathString);
+    if (path.toString().endsWith(".zip")) {
```
[the remark this answers is not in the document]
Comment (cyulysses-corp) [reply src/java/com/google/cloud/bigquery/dwhassessment/extractiontool/subcommand/ExtractSubcommand.java]: Done.

File: src/java/com/google/cloud/bigquery/dwhassessment/extractiontool/subcommand/ExtractSubcommand.java [hunk 8]
```diff
@@ -408,6 +389,64 @@ private ExtractExecutor.Arguments getValidatedArguments() {
     return arguments;
   }
 
+  private void setOutputPath() {
```
[the remark this answers is not in the document]
Comment (cyulysses-corp) [reply src/java/com/google/cloud/bigquery/dwhassessment/extractiontool/subcommand/ExtractSubcommand.java]: Renamed.

File: src/java/com/google/cloud/bigquery/dwhassessment/extractiontool/executor/SaveCheckerImpl.java [hunk 9]
```diff
@@ -138,6 +140,17 @@ public ImmutableMap<String, ChunkCheckpoint> getScriptCheckPoints(Path path) {
     return checkPointsMapBuilder.build();
   }
 
+  @Override
+  public ImmutableSet<String> findScriptNamesWithFinishedRecords(
+      Path recordPath, Set<String> scriptsToCheck, String fileSuffix) {
+    if (scriptsToCheck == null || scriptsToCheck.isEmpty()) {
```
[the remark this answers is not in the document]
Comment (cyulysses-corp) [reply src/java/com/google/cloud/bigquery/dwhassessment/extractiontool/executor/SaveCheckerImpl.java]: Removed the null check.

File: src/java/com/google/cloud/bigquery/dwhassessment/extractiontool/executor/SaveChecker.java [hunk 10]
```diff
@@ -30,4 +32,17 @@ public interface SaveChecker {
...
   ImmutableMap<String, ChunkCheckpoint> getScriptCheckPoints(Path path);
+
+  /**
+   * Returns a set containing the names of the scripts whose corresponding finished records have
+   * been found in the target path. It usually means that these scripts have finished successfully
+   * during a previous run.
+   *
+   * @param recordPath Target path containing records from the previous run(s).
+   * @param scriptsToCheck The scripts whose records to look for.
+   * @param fileSuffix The ending string of the record files.
```
[the remark this answers is not in the document]
Comment (cyulysses-corp) [reply src/java/com/google/cloud/bigquery/dwhassessment/extractiontool/executor/SaveChecker.java]: Changed it to "extension"

File: src/java/com/google/cloud/bigquery/dwhassessment/extractiontool/executor/ExtractExecutorImpl.java [hunk 11]
```diff
@@ -126,19 +126,38 @@ private static void maybeAddTimeRange(
...
+        arguments.mode().equals(RunMode.NORMAL) || arguments.prevRunPath().isPresent(),
+        "Value prevRunPath is not defined while the mode is not NORMAL; this should not happen.");
+
     DataEntityManager dataEntityManager = dataEntityManagerFactory.apply(arguments.outputPath());
+
+    // Determine the scripts to run.
+    ImmutableSet<String> requestedScripts = getRequestedScripts(arguments);
+    if (arguments.mode() == RunMode.RECOVERY) {
+      ImmutableSet<String> finishedScripts =
```
[the remark this answers is not in the document]
Comment (cyulysses-corp) [reply src/java/com/google/cloud/bigquery/dwhassessment/extractiontool/executor/ExtractExecutorImpl.java]: Gave some thoughts to it and:
* Now logging both finished scripts and scripts that will be run.
* Did not change the SaveChecker's method's behaviour:
  * It is not SaveChecker's duty to decide which scripts to run; it should only report the saves but the decision should happen within the executor.
  * Changed the method name to make it clearer.
  * Consolidated the decision logic into "getRequestedScripts".

File: src/javatests/com/google/cloud/bigquery/dwhassessment/extractiontool/executor/SaveCheckerImplTest.java (hunk 1 above)
Comment (cyulysses-corp) [reply src/javatests/com/google/cloud/bigquery/dwhassessment/extractiontool/executor/SaveCheckerImplTest.java]: Added tests for the first; the latter two cases aren't qualitatively different from the canonical workflow.

File: src/java/com/google/cloud/bigquery/dwhassessment/extractiontool/executor/ExtractExecutorImpl.java (hunk 3 above)
Comment (cyulysses-corp) [reply src/java/com/google/cloud/bigquery/dwhassessment/extractiontool/executor/ExtractExecutorImpl.java:129]: Done.

File: src/java/com/google/cloud/bigquery/dwhassessment/extractiontool/subcommand/ExtractSubcommand.java (hunk 5 above)
Comment (cyulysses-corp) [reply src/java/com/google/cloud/bigquery/dwhassessment/extractiontool/subcommand/ExtractSubcommand.java:443]: Done.

File: src/java/com/google/cloud/bigquery/dwhassessment/extractiontool/subcommand/ExtractSubcommand.java (hunk 6 above)
Comment (cyulysses-corp) [reply src/java/com/google/cloud/bigquery/dwhassessment/extractiontool/subcommand/ExtractSubcommand.java:434]: Done.

File: src/javatests/com/google/cloud/bigquery/dwhassessment/extractiontool/executor/ExtractExecutorImplTest.java (hunk 2 above)
Comment (cyulysses-corp) [reply src/javatests/com/google/cloud/bigquery/dwhassessment/extractiontool/executor/ExtractExecutorImplTest.java]: Actually for the "no chunk" test we do want to verify the chunk-related method is not called (i.e. RECOVERY mode is decoupled from chunking requirements). For the "chunk" test it is indeed true and I removed verifyNoMoreInteractions checks.

File: src/java/com/google/cloud/bigquery/dwhassessment/extractiontool/subcommand/ExtractSubcommand.java (hunk 4 above)
Comment (cyulysses-corp) [reply src/java/com/google/cloud/bigquery/dwhassessment/extractiontool/subcommand/ExtractSubcommand.java:348]: Makes sense. I also think the naming of this class itself is a bit confusing. Will do the housekeeping work in another CL.

File: src/java/com/google/cloud/bigquery/dwhassessment/extractiontool/executor/ExtractExecutorImpl.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (frsv) [reply src/java/com/google/cloud/bigquery/dwhassessment/extractiontool/executor/ExtractExecutorImpl.java]: Thank you!

File: src/java/com/google/cloud/bigquery/dwhassessment/extractiontool/executor/SaveChecker.java [hunk 12]
```diff
@@ -30,4 +32,17 @@ public interface SaveChecker {
    * @return The <scriptName, chunkCheckPoint> map indicating where the new run should start from.
    */
   ImmutableMap<String, ChunkCheckpoint> getScriptCheckPoints(Path path);
+
+  /**
+   * Given a set of scripts, returns the names of those that have finished successfully during a
+   * previous run. The "finishedness" is inferred from the names of the files present in the target
```
Comment (frsv) [inline src/java/com/google/cloud/bigquery/dwhassessment/extractiontool/executor/SaveChecker.java:38]: nit: completed / completeness sounds more natural

**Review by frsv — APPROVED** [review]

File: src/java/com/google/cloud/bigquery/dwhassessment/extractiontool/executor/SaveChecker.java (hunk 12 above)
Comment (cyulysses-corp) [reply src/java/com/google/cloud/bigquery/dwhassessment/extractiontool/executor/SaveChecker.java]: Good call, although complete/completeness may have overloaded meanings in the context of data chunking. I rephrased it to comply more with the teachings of tech writers. :-)
